### ISSUE TITLE
About the threshold of audio tagging using DESED dataset

### ISSUE THREAD

**Issue by SteveTanggithub** [RicherMans/UIT_Mobile#2]
I want to use DESED dataset which is part of the audioset for audio tagging, the DESED consists of 10 classes events. How should I set the threshold when outputting the pred result? Why do u set 0.2 as ur threshold? Thank u so much!

**Comment by RicherMans** [comment]
Hey @SteveTanggithub ,
so generally speaking, if you want to predict hard labels you can choose any threshold. The performance will greatly differ between most thresholds of say 0 to 1.

Please note that for this work, only the keyword spotting branch is thresholded,  while the audio tagging branch just predicts raw scores.
This is because Audioset uses a non-threshold evaluation metric by default ( mean average precision ).

Why did we choose 0.2 in this paper? Because we simply went once over the dev set on GSC and found that 0.2 works best to "optimize for accuracy".
In real-world applications, however, I would not choose 0.2, but rather 0.5 or higher, since they change of obtaining a false-positive is high when using low thresholds.
All in all you can understand the choice of threshold as a necessity to obtain a   "publishable result", which one would not implement for other datasets.

**Comment by SteveTanggithub** [comment]
my true label is like [1,0,0,1,0,0,0,0,1,0], its multi-hot label. How do I decide which position of the UIT model should be 1 or 0 if I DONT use the threshold? And how can I calculate the metric like recall,acc, or precision? I am not sure I make my question comprehensible. Looking forward to your answer :) By the way, what is the groundtruth(true label) of Audioset?

**Comment by RicherMans** [comment]
Hey, 
so I understand your point here.
Let me explain it this way:
There are two main directions for sound events, one called Sound Event Detection (SED) and the other being Audio Tagging (AT).
Both aim to predict sound event labels from a given audio clip.
However, SED works like that: You provide it a sound clip of some length, say 30s, and it detects you each sound events *plus* start and end points of each sound event in the clip. The entire process is generally **offline**, i.e., you give the model first 30s of data, then it predicts the results.

AT only predicts if a label is occuring [0,1]. It does not answer the question of "where was the event". AT can be done **online**, meaning you can feed the model maybe shorter chunks say 1s in order to get some results on a fine scale.

This work here only focuses on AT.
Further, this work is not focusing on performance, we focus on real-world deployment.

> And how can I calculate the metric like recall,acc, or precision?

I hope that helps: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html 

Again in this work, we use Accuracy **for KWS** and **mean average precision** (mAP, link above) for Tagging.

**Comment by SteveTanggithub** [comment]
and i have some more questions:
1.the audioset has 527 event classes u used, what will mAP be in normal situation if i use the DESED dataset that has 10 classes? I have trained UIT_XXXS model using DESED but my mAP is only about 0.46.  
2.how can i get the true event label of 1s audio which is part of 10s audio clip?
3.most methods in SED use mean-teacher training skill for utilizing plenty of unlabeled data, is it necessary for using mean-teacher training in audio tagging?
Thank you so much:)

**Comment by RicherMans** [comment]
Hey,
> 1.the audioset has 527 event classes u used, what will mAP be in normal situation if i use the DESED dataset that has 10 classes? I have trained UIT_XXXS model using DESED but my mAP is only about 0.46.

Well I would expect that the mAP would be a bit higher, but the model itself is just optimized for inference not to obtain the best and highest results possible.
Also we trained using pseudo-strong-labels (scale of 1s on Audioset), which makes a big difference. 

> 2.how can i get the true event label of 1s audio which is part of 10s audio clip?

Well just use another model and predict it, otherwise there is no way. 

> 3.most methods in SED use mean-teacher training skill for utilizing plenty of unlabeled data, is it necessary for using mean-teacher training in audio tagging?

Actually for DESED , mean teacher only works because the so-called "unlabeled" dataset contains your target labels (target domain, the 10 classes) and is not truly "unlabeled". The dataset before 2020, had an unlabeled out-of-domain subset. In my experience,  using that out-of-domain dataset always decreases performance when using mean-teacher, because the teacher labels are simply incorrect.
In fact, I prefer using Unsupervised data augmentation (UDA) on DESED and never see any huge difference between the methods.
What I'm trying to say is, I guess mean teacher is slightly better than just using your model to predict the labels directly, but not by a noticeable margin.

A long time ago I used mean-teacher on audioset, but the results are the same: It only works for some small-scale datasets, with some strict conditioning on unlabeled data.
